clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -149,6 +157,74 @@ public Sender(LogContext logContext,
         this.retryBackoffMs = retryBackoffMs;
         this.apiVersions = apiVersions;
         this.transactionManager = transactionManager;
+        this.inFlightBatches = new HashMap<>();
+    }
+
+    public Map<TopicPartition, List<ProducerBatch>> getInflightBatches() {
+        return inFlightBatches;
+    }
+
+    public List<ProducerBatch> inFlightBatches(TopicPartition tp) {
+        return inFlightBatches.containsKey(tp) ? inFlightBatches.get(tp) : new ArrayList<>();
+    }
+
+    public void maybeRemoveFromInflightBatches(ProducerBatch batch) {
+        List<ProducerBatch> batches = inFlightBatches.get(batch.topicPartition);
+        if (batches != null) {
```

Comment:
If `batches` is empty, should we remove the list from `inFlightBatches`?

Reply:
updated the change to remove the list from `inFlightBatches` when it is empty.